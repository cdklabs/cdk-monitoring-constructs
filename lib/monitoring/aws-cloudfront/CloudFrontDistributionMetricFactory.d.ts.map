{"version":3,"file":"CloudFrontDistributionMetricFactory.d.ts","sourceRoot":"","sources":["CloudFrontDistributionMetricFactory.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,aAAa,EAAE,MAAM,4BAA4B,CAAC;AAG3D,OAAO,EACL,aAAa,EAEb,qBAAqB,EACtB,MAAM,cAAc,CAAC;AAMtB,MAAM,WAAW,wCAAwC;IACvD,QAAQ,CAAC,YAAY,EAAE,aAAa,CAAC;IACrC;;OAEG;IACH,QAAQ,CAAC,iBAAiB,CAAC,EAAE,OAAO,CAAC;IACrC;;OAEG;IACH,QAAQ,CAAC,qBAAqB,CAAC,EAAE,qBAAqB,CAAC;CACxD;AAED;;;GAGG;AACH,qBAAa,mCAAmC;IAC9C,OAAO,CAAC,QAAQ,CAAC,aAAa,CAAgB;IAC9C,OAAO,CAAC,QAAQ,CAAC,iBAAiB,CAAU;IAC5C,OAAO,CAAC,QAAQ,CAAC,qBAAqB,CAAwB;IAC9D,OAAO,CAAC,QAAQ,CAAC,aAAa,CAAgB;gBAG5C,aAAa,EAAE,aAAa,EAC5B,KAAK,EAAE,wCAAwC;IAYjD,kBAAkB;IAalB,iBAAiB;IAUjB;;OAEG;IACH,gBAAgB;IAUhB,wBAAwB;IAaxB,0BAA0B;IAa1B,kCAAkC;IAalC,yBAAyB;IAazB,yBAAyB;IAazB,2BAA2B;CAY5B"}