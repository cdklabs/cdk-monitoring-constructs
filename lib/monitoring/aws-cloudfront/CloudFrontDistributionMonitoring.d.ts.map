{"version":3,"file":"CloudFrontDistributionMonitoring.d.ts","sourceRoot":"","sources":["CloudFrontDistributionMonitoring.ts"],"names":[],"mappings":"AAAA,OAAO,EACL,WAAW,EACX,oBAAoB,EACpB,OAAO,EACR,MAAM,4BAA4B,CAAC;AAEpC,OAAO,EACL,YAAY,EACZ,mBAAmB,EAGnB,iBAAiB,EACjB,kBAAkB,EAGlB,gBAAgB,EAChB,eAAe,EACf,sBAAsB,EACtB,UAAU,EACV,eAAe,EAKf,eAAe,EAChB,MAAM,cAAc,CAAC;AACtB,OAAO,EACL,sBAAsB,EACtB,wBAAwB,EACzB,MAAM,iBAAiB,CAAC;AACzB,OAAO,EAEL,wCAAwC,EACzC,MAAM,uCAAuC,CAAC;AAE/C,MAAM,WAAW,uCACf,SAAQ,mBAAmB;CAAG;AAEhC,MAAM,WAAW,qCACf,SAAQ,wCAAwC,EAC9C,uCAAuC;IACzC,QAAQ,CAAC,cAAc,CAAC,EAAE,MAAM,CAAC,MAAM,EAAE,eAAe,CAAC,CAAC;IAC1D,QAAQ,CAAC,eAAe,CAAC,EAAE,MAAM,CAAC,MAAM,EAAE,gBAAgB,CAAC,CAAC;IAC5D,QAAQ,CAAC,eAAe,CAAC,EAAE,MAAM,CAAC,MAAM,EAAE,kBAAkB,CAAC,CAAC;IAC9D,QAAQ,CAAC,eAAe,CAAC,EAAE,MAAM,CAAC,MAAM,EAAE,kBAAkB,CAAC,CAAC;CAC/D;AAED,qBAAa,gCAAiC,SAAQ,UAAU;IAC9D,OAAO,CAAC,QAAQ,CAAC,KAAK,CAAS;IAC/B,OAAO,CAAC,QAAQ,CAAC,eAAe,CAAC,CAAS;IAE1C,SAAS,CAAC,QAAQ,CAAC,cAAc,EAAE,wBAAwB,CAAC;IAC5D,SAAS,CAAC,QAAQ,CAAC,YAAY,EAAE,YAAY,CAAC;IAC9C,SAAS,CAAC,QAAQ,CAAC,iBAAiB,EAAE,iBAAiB,CAAC;IACxD,SAAS,CAAC,QAAQ,CAAC,eAAe,EAAE,eAAe,CAAC;IAEpD,SAAS,CAAC,QAAQ,CAAC,oBAAoB,EAAE,oBAAoB,EAAE,CAAC;IAChE,SAAS,CAAC,QAAQ,CAAC,cAAc,EAAE,oBAAoB,EAAE,CAAC;IAE1D,SAAS,CAAC,QAAQ,CAAC,SAAS,EAAE,sBAAsB,CAAC;IACrD,SAAS,CAAC,QAAQ,CAAC,qBAAqB,EAAE,sBAAsB,CAAC;IACjE,SAAS,CAAC,QAAQ,CAAC,mBAAmB,EAAE,sBAAsB,CAAC;IAC/D,SAAS,CAAC,QAAQ,CAAC,YAAY,EAAE,sBAAsB,CAAC;IACxD,SAAS,CAAC,QAAQ,CAAC,YAAY,EAAE,sBAAsB,CAAC;IACxD,SAAS,CAAC,QAAQ,CAAC,YAAY,EAAE,sBAAsB,CAAC;gBAGtD,KAAK,EAAE,eAAe,EACtB,KAAK,EAAE,qCAAqC;IAqF9C,cAAc,IAAI,OAAO,EAAE;IAQ3B,OAAO,IAAI,OAAO,EAAE;IAUpB,SAAS,CAAC,iBAAiB;IAQ3B,SAAS,CAAC,eAAe,CAAC,KAAK,EAAE,MAAM,EAAE,MAAM,EAAE,MAAM;IAWvD,SAAS,CAAC,iBAAiB,CAAC,KAAK,EAAE,MAAM,EAAE,MAAM,EAAE,MAAM;IAUzD,SAAS,CAAC,mBAAmB,CAAC,KAAK,EAAE,MAAM,EAAE,MAAM,EAAE,MAAM;IAU3D,SAAS,CAAC,qBAAqB,CAAC,KAAK,EAAE,MAAM,EAAE,MAAM,EAAE,MAAM;CAU9D"}