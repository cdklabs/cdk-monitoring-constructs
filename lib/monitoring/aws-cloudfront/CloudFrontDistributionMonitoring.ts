import {
  GraphWidget,
  HorizontalAnnotation,
  IWidget,
} from "aws-cdk-lib/aws-cloudwatch";

import {
  CloudFrontDistributionMetricFactory,
  CloudFrontDistributionMetricFactoryProps,
} from "./CloudFrontDistributionMetricFactory";
import {
  AlarmFactory,
  BaseMonitoringProps,
  DefaultGraphWidgetHeight,
  DefaultSummaryWidgetHeight,
  ErrorAlarmFactory,
  ErrorRateThreshold,
  ErrorType,
  HalfWidth,
  HighTpsThreshold,
  LowTpsThreshold,
  MetricWithAlarmSupport,
  Monitoring,
  MonitoringScope,
  PercentageAxisFromZeroToHundred,
  QuarterWidth,
  RateAxisFromZero,
  SizeAxisBytesFromZero,
  ThirdWidth,
  TpsAlarmFactory,
} from "../../common";
import {
  MonitoringHeaderWidget,
  MonitoringNamingStrategy,
} from "../../dashboard";

export interface CloudFrontDistributionMonitoringOptions
  extends BaseMonitoringProps {}

export interface CloudFrontDistributionMonitoringProps
  extends CloudFrontDistributionMetricFactoryProps,
    CloudFrontDistributionMonitoringOptions {
  readonly addError4xxRate?: Record<string, ErrorRateThreshold>;
  readonly addFault5xxRate?: Record<string, ErrorRateThreshold>;

  readonly addLowTpsAlarm?: Record<string, LowTpsThreshold>;
  readonly addHighTpsAlarm?: Record<string, HighTpsThreshold>;
}

export class CloudFrontDistributionMonitoring extends Monitoring {
  readonly title: string;
  readonly distributionUrl?: string;

  readonly namingStrategy: MonitoringNamingStrategy;
  readonly alarmFactory: AlarmFactory;
  readonly errorAlarmFactory: ErrorAlarmFactory;
  readonly tpsAlarmFactory: TpsAlarmFactory;

  readonly errorRateAnnotations: HorizontalAnnotation[];
  readonly tpsAnnotations: HorizontalAnnotation[];

  readonly tpsMetric: MetricWithAlarmSupport;
  readonly downloadedBytesMetric: MetricWithAlarmSupport;
  readonly uploadedBytesMetric: MetricWithAlarmSupport;
  readonly error4xxRate: MetricWithAlarmSupport;
  readonly error5xxRate: MetricWithAlarmSupport;

  readonly additionalMetricsEnabled: boolean;
  readonly cacheHitRate: MetricWithAlarmSupport | undefined;

  constructor(
    scope: MonitoringScope,
    props: CloudFrontDistributionMonitoringProps
  ) {
    super(scope);

    const namedConstruct = props.distribution;
    const fallbackConstructName = namedConstruct.distributionId;

    this.namingStrategy = new MonitoringNamingStrategy({
      ...props,
      namedConstruct,
      fallbackConstructName,
    });
    this.title = this.namingStrategy.resolveHumanReadableName();
    this.distributionUrl = scope
      .createAwsConsoleUrlFactory()
      .getCloudFrontDistributionUrl(namedConstruct.distributionId);

    this.alarmFactory = this.createAlarmFactory(
      this.namingStrategy.resolveAlarmFriendlyName()
    );
    this.errorAlarmFactory = new ErrorAlarmFactory(this.alarmFactory);
    this.tpsAlarmFactory = new TpsAlarmFactory(this.alarmFactory);

    this.errorRateAnnotations = [];
    this.tpsAnnotations = [];

    const metricFactory = new CloudFrontDistributionMetricFactory(
      scope.createMetricFactory(),
      props
    );
    this.tpsMetric = metricFactory.metricRequestTps();
    this.downloadedBytesMetric = metricFactory.metricTotalBytesDownloaded();
    this.uploadedBytesMetric = metricFactory.metricTotalBytesUploaded();
    this.error4xxRate = metricFactory.metric4xxErrorRateAverage();
    this.error5xxRate = metricFactory.metric5xxErrorRateAverage();

    this.additionalMetricsEnabled = props.additionalMetricsEnabled ?? true;
    if (this.additionalMetricsEnabled) {
      this.cacheHitRate = metricFactory.metricCacheHitRateAverageInPercent();
    }

    for (const disambiguator in props.addLowTpsAlarm) {
      const alarmProps = props.addLowTpsAlarm[disambiguator];
      const createdAlarm = this.tpsAlarmFactory.addMinTpsAlarm(
        this.tpsMetric,
        alarmProps,
        disambiguator
      );
      this.tpsAnnotations.push(createdAlarm.annotation);
      this.addAlarm(createdAlarm);
    }
    for (const disambiguator in props.addHighTpsAlarm) {
      const alarmProps = props.addHighTpsAlarm[disambiguator];
      const createdAlarm = this.tpsAlarmFactory.addMaxTpsAlarm(
        this.tpsMetric,
        alarmProps,
        disambiguator
      );
      this.tpsAnnotations.push(createdAlarm.annotation);
      this.addAlarm(createdAlarm);
    }
    for (const disambiguator in props.addError4xxRate) {
      const alarmProps = props.addError4xxRate[disambiguator];
      const createdAlarm = this.errorAlarmFactory.addErrorRateAlarm(
        this.error4xxRate,
        ErrorType.ERROR,
        alarmProps,
        disambiguator
      );
      this.errorRateAnnotations.push(createdAlarm.annotation);
      this.addAlarm(createdAlarm);
    }
    for (const disambiguator in props.addFault5xxRate) {
      const alarmProps = props.addFault5xxRate[disambiguator];
      const createdAlarm = this.errorAlarmFactory.addErrorRateAlarm(
        this.error5xxRate,
        ErrorType.FAULT,
        alarmProps,
        disambiguator
      );
      this.errorRateAnnotations.push(createdAlarm.annotation);
      this.addAlarm(createdAlarm);
    }

    if (props.useCreatedAlarms) {
      props.useCreatedAlarms.consume(this.createdAlarms());
    }
  }

  summaryWidgets(): IWidget[] {
    return [
      this.createTitleWidget(),
      this.createTpsWidget(HalfWidth, DefaultSummaryWidgetHeight),
      this.createErrorRateWidget(HalfWidth, DefaultSummaryWidgetHeight),
    ];
  }

  widgets(): IWidget[] {
    if (this.additionalMetricsEnabled) {
      return [
        this.createTitleWidget(),
        this.createTpsWidget(QuarterWidth, DefaultGraphWidgetHeight),
        this.createCacheWidget(QuarterWidth, DefaultGraphWidgetHeight),
        this.createTrafficWidget(QuarterWidth, DefaultGraphWidgetHeight),
        this.createErrorRateWidget(QuarterWidth, DefaultGraphWidgetHeight),
      ];
    } else {
      return [
        this.createTitleWidget(),
        this.createTpsWidget(ThirdWidth, DefaultGraphWidgetHeight),
        this.createTrafficWidget(ThirdWidth, DefaultGraphWidgetHeight),
        this.createErrorRateWidget(ThirdWidth, DefaultGraphWidgetHeight),
      ];
    }
  }

  createTitleWidget() {
    return new MonitoringHeaderWidget({
      family: "CloudFront Distribution",
      title: this.title,
      goToLinkUrl: this.distributionUrl,
    });
  }

  createTpsWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: "TPS",
      left: [this.tpsMetric],
      leftYAxis: RateAxisFromZero,
      leftAnnotations: this.tpsAnnotations,
    });
  }

  createCacheWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: "Hit Rate",
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      left: [this.cacheHitRate!],
      leftYAxis: PercentageAxisFromZeroToHundred,
    });
  }

  createTrafficWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: "Traffic",
      left: [this.downloadedBytesMetric, this.uploadedBytesMetric],
      leftYAxis: SizeAxisBytesFromZero,
    });
  }

  createErrorRateWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: "Errors (rate)",
      left: [this.error4xxRate, this.error5xxRate],
      leftAnnotations: this.errorRateAnnotations,
      leftYAxis: RateAxisFromZero,
    });
  }
}
