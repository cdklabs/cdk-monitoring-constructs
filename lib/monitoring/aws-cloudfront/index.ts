export * from "./CloudFrontDistributionMetricFactory";
export * from "./CloudFrontDistributionMonitoring";
