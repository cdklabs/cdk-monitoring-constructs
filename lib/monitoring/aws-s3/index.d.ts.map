{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["index.ts"],"names":[],"mappings":"AAAA,cAAc,yBAAyB,CAAC;AACxC,cAAc,sBAAsB,CAAC"}