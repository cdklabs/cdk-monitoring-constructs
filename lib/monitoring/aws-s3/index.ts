export * from "./S3BucketMetricFactory";
export * from "./S3BucketMonitoring";
