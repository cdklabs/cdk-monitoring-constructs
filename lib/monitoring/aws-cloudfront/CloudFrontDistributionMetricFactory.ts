import { IDistribution } from "aws-cdk-lib/aws-cloudfront";
import { DimensionsMap } from "aws-cdk-lib/aws-cloudwatch";

import {
  MetricFactory,
  MetricStatistic,
  RateComputationMethod,
} from "../../common";

const CloudFrontNamespace = "AWS/CloudFront";
const CloudFrontGlobalRegion = "Global";
const CloudFrontDefaultMetricRegion = "us-east-1";

export interface CloudFrontDistributionMetricFactoryProps {
  readonly distribution: IDistribution;

  /**
   * @default - true
   */
  readonly fillTpsWithZeroes?: boolean;

  /**
   * @default - average
   */
  readonly rateComputationMethod?: RateComputationMethod;

  /**
   * Generate dashboard charts for additional CloudFront distribution metrics.
   *
   * To enable additional metrics on your CloudFront distribution, see
   * https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/viewing-cloudfront-metrics.html#monitoring-console.distributions-additional
   *
   * @default - true
   */
  readonly additionalMetricsEnabled?: boolean;
}

/**
 * To get the CloudFront metrics from the CloudWatch API, you must use the US East (N. Virginia) Region (us-east-1).
 * https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/programming-cloudwatch-metrics.html
 */
export class CloudFrontDistributionMetricFactory {
  private readonly metricFactory: MetricFactory;
  private readonly fillTpsWithZeroes: boolean;
  private readonly rateComputationMethod: RateComputationMethod;
  private readonly dimensionsMap: DimensionsMap;

  constructor(
    metricFactory: MetricFactory,
    props: CloudFrontDistributionMetricFactoryProps
  ) {
    this.metricFactory = metricFactory;
    this.fillTpsWithZeroes = props.fillTpsWithZeroes ?? true;
    this.rateComputationMethod =
      props.rateComputationMethod ?? RateComputationMethod.AVERAGE;
    this.dimensionsMap = {
      DistributionId: props.distribution.distributionId,
      Region: CloudFrontGlobalRegion,
    };
  }

  metricRequestCount() {
    return this.metricFactory.createMetric(
      "Requests",
      MetricStatistic.SUM,
      "Uploaded",
      this.dimensionsMap,
      undefined,
      CloudFrontNamespace,
      undefined,
      CloudFrontDefaultMetricRegion
    );
  }

  metricRequestRate() {
    return this.metricFactory.toRate(
      this.metricRequestCount(),
      this.rateComputationMethod,
      false,
      "requests",
      this.fillTpsWithZeroes
    );
  }

  /**
   * @deprecated use metricRequestRate
   */
  metricRequestTps() {
    return this.metricFactory.toRate(
      this.metricRequestCount(),
      RateComputationMethod.PER_SECOND,
      false,
      "requests",
      this.fillTpsWithZeroes
    );
  }

  metricTotalBytesUploaded() {
    return this.metricFactory.createMetric(
      "BytesUploaded",
      MetricStatistic.SUM,
      "Uploaded",
      this.dimensionsMap,
      undefined,
      CloudFrontNamespace,
      undefined,
      CloudFrontDefaultMetricRegion
    );
  }

  metricTotalBytesDownloaded() {
    return this.metricFactory.createMetric(
      "BytesDownloaded",
      MetricStatistic.SUM,
      "Downloaded",
      this.dimensionsMap,
      undefined,
      CloudFrontNamespace,
      undefined,
      CloudFrontDefaultMetricRegion
    );
  }

  /**
   * Cache hit rate metric. This is an additional metric that needs to be explicitly enabled for an additional cost.
   *
   * @see https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/viewing-cloudfront-metrics.html#monitoring-console.distributions-additional
   */
  metricCacheHitRateAverageInPercent() {
    return this.metricFactory.createMetric(
      "CacheHitRate",
      MetricStatistic.AVERAGE,
      "Hit Rate",
      this.dimensionsMap,
      undefined,
      CloudFrontNamespace,
      undefined,
      CloudFrontDefaultMetricRegion
    );
  }

  metric4xxErrorRateAverage() {
    return this.metricFactory.createMetric(
      "4xxErrorRate",
      MetricStatistic.AVERAGE,
      "4XX",
      this.dimensionsMap,
      undefined,
      CloudFrontNamespace,
      undefined,
      CloudFrontDefaultMetricRegion
    );
  }

  metric5xxErrorRateAverage() {
    return this.metricFactory.createMetric(
      "5xxErrorRate",
      MetricStatistic.AVERAGE,
      "5XX",
      this.dimensionsMap,
      undefined,
      CloudFrontNamespace,
      undefined,
      CloudFrontDefaultMetricRegion
    );
  }

  metricTotalErrorRateAverage() {
    return this.metricFactory.createMetric(
      "TotalErrorRate",
      MetricStatistic.AVERAGE,
      "Total",
      this.dimensionsMap,
      undefined,
      CloudFrontNamespace,
      undefined,
      CloudFrontDefaultMetricRegion
    );
  }
}
