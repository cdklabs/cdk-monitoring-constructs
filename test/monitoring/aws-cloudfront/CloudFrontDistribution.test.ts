import { Stack } from "monocdk";
import { Template } from "monocdk/assertions";
import { Distribution } from "monocdk/aws-cloudfront";
import { S3Origin } from "monocdk/aws-cloudfront-origins";
import { Bucket } from "monocdk/aws-s3";

import {
  AlarmWithAnnotation,
  CloudFrontDistributionMonitoring,
} from "../../../lib";
import { TestMonitoringScope } from "../TestMonitoringScope";

test("snapshot test: no alarms", () => {
  const stack = new Stack();
  const bucket = new Bucket(stack, "Bucket");
  const distribution = new Distribution(stack, "Distribution", {
    defaultBehavior: {
      origin: new S3Origin(bucket),
    },
  });

  const scope = new TestMonitoringScope(stack, "Scope");

  new CloudFrontDistributionMonitoring(scope, { distribution });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});

test("snapshot test: all alarms", () => {
  const stack = new Stack();
  const bucket = new Bucket(stack, "Bucket");
  const distribution = new Distribution(stack, "Distribution", {
    defaultBehavior: {
      origin: new S3Origin(bucket),
    },
  });

  const scope = new TestMonitoringScope(stack, "Scope");

  let numAlarmsCreated = 0;

  new CloudFrontDistributionMonitoring(scope, {
    distribution,
    addLowTpsAlarm: {
      Warning: {
        minTps: 10,
      },
    },
    addHighTpsAlarm: {
      Warning: {
        maxTps: 20,
      },
    },
    addError4xxRate: {
      Warning: {
        maxErrorRate: 0.5,
      },
    },
    addFault5xxRate: {
      Warning: {
        maxErrorRate: 0.8,
      },
    },
    useCreatedAlarms: {
      consume(alarms: AlarmWithAnnotation[]) {
        numAlarmsCreated = alarms.length;
      },
    },
  });

  expect(numAlarmsCreated).toStrictEqual(4);
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
