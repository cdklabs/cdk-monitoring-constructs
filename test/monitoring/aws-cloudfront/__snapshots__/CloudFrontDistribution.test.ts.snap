// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test: all alarms 1`] = `
Object {
  "Resources": Object {
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketPolicyE9A3008A": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "DistributionOrigin1S3Origin5F5C0696",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "Distribution830FAC52": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "DefaultCacheBehavior": Object {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "DistributionOrigin13547B94F",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "Bucket83908E77",
                  "RegionalDomainName",
                ],
              },
              "Id": "DistributionOrigin13547B94F",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "DistributionOrigin1S3Origin5F5C0696",
                      },
                    ],
                  ],
                },
              },
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "DistributionOrigin1S3Origin5F5C0696": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Identity for DistributionOrigin13547B94F",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "ScopeTestDistributionErrorRateWarning079D566A": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Error rate is too high.",
        "AlarmName": "Test-Distribution-Error-Rate-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "4XX",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "DistributionId",
                    "Value": Object {
                      "Ref": "Distribution830FAC52",
                    },
                  },
                  Object {
                    "Name": "Region",
                    "Value": "Global",
                  },
                ],
                "MetricName": "4xxErrorRate",
                "Namespace": "AWS/CloudFront",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 0.5,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDistributionFaultRateWarning02013C05": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "Fault rate is too high.",
        "AlarmName": "Test-Distribution-Fault-Rate-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Id": "m1",
            "Label": "5XX",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "DistributionId",
                    "Value": Object {
                      "Ref": "Distribution830FAC52",
                    },
                  },
                  Object {
                    "Name": "Region",
                    "Value": "Global",
                  },
                ],
                "MetricName": "5xxErrorRate",
                "Namespace": "AWS/CloudFront",
              },
              "Period": 300,
              "Stat": "Average",
            },
            "ReturnData": true,
          },
        ],
        "Threshold": 0.8,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDistributionMaxTPSWarningAC289EB2": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "TPS is too high.",
        "AlarmName": "Test-Distribution-MaxTPS-Warning",
        "ComparisonOperator": "GreaterThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Expression": "FILL(requests,0) / PERIOD(requests)",
            "Id": "expr_1",
            "Label": "Uploaded/s",
          },
          Object {
            "Id": "requests",
            "Label": "Uploaded",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "DistributionId",
                    "Value": Object {
                      "Ref": "Distribution830FAC52",
                    },
                  },
                  Object {
                    "Name": "Region",
                    "Value": "Global",
                  },
                ],
                "MetricName": "Requests",
                "Namespace": "AWS/CloudFront",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 20,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ScopeTestDistributionMinTPSWarning6CB7ACEE": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmDescription": "TPS is too low.",
        "AlarmName": "Test-Distribution-MinTPS-Warning",
        "ComparisonOperator": "LessThanThreshold",
        "DatapointsToAlarm": 3,
        "EvaluationPeriods": 3,
        "Metrics": Array [
          Object {
            "Expression": "FILL(requests,0) / PERIOD(requests)",
            "Id": "expr_1",
            "Label": "Uploaded/s",
          },
          Object {
            "Id": "requests",
            "Label": "Uploaded",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "DistributionId",
                    "Value": Object {
                      "Ref": "Distribution830FAC52",
                    },
                  },
                  Object {
                    "Name": "Region",
                    "Value": "Global",
                  },
                ],
                "MetricName": "Requests",
                "Namespace": "AWS/CloudFront",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 10,
        "TreatMissingData": "missing",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
}
`;

exports[`snapshot test: no alarms 1`] = `
Object {
  "Resources": Object {
    "Bucket83908E77": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "BucketPolicyE9A3008A": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "Bucket83908E77",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "DistributionOrigin1S3Origin5F5C0696",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "Bucket83908E77",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "Distribution830FAC52": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "DefaultCacheBehavior": Object {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "DistributionOrigin13547B94F",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "Bucket83908E77",
                  "RegionalDomainName",
                ],
              },
              "Id": "DistributionOrigin13547B94F",
              "S3OriginConfig": Object {
                "OriginAccessIdentity": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "origin-access-identity/cloudfront/",
                      Object {
                        "Ref": "DistributionOrigin1S3Origin5F5C0696",
                      },
                    ],
                  ],
                },
              },
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "DistributionOrigin1S3Origin5F5C0696": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": "Identity for DistributionOrigin13547B94F",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
  },
}
`;
